### PR TITLE
Kernel injection fixes

### PR DESCRIPTION
**Pull request by tjruwase**
Add replacement API with configurable kernel optimizations
Release context memory ASAP
Add training flag to replacement API

### CODE REVIEW

File: deepspeed/ops/transformer/transformer.py [hunk 1]
```diff
@@ -405,6 +408,25 @@ def backward(ctx, grad_output):
              norm_w,
              norm_b)
 
+        # This appears to be an effective way to release context memory
```
Comment (RezaYazdaniAminabadi) [inline deepspeed/ops/transformer/transformer.py:411]: Did this have any performance implication at all!?

File: deepspeed/ops/transformer/transformer.py [hunk 2]
```diff
@@ -248,7 +251,7 @@ def forward(ctx,
             norm_w.register_hook(lambda x, self=self: grads.append([x, "norm_W"]))
             norm_b.register_hook(lambda x, self=self: grads.append([x, "norm_B"]))
 
-        if config.is_grad_enabled:
+        if config.is_grad_enabled and config.training:
```
Comment (RezaYazdaniAminabadi) [inline deepspeed/ops/transformer/transformer.py:254]: this is so useful checking! :)

**Review by RezaYazdaniAminabadi — APPROVED** [review]

**Comment by jeffra** [comment]
Offline discussion, new API to implement:

```
config = deepspeed.ops.DeepSpeedTransformerConfig(bsz=32, seq=128, ..., huggingface=true)
model = deepspeed.ops.replace_module(
           orig_module_impl=transformers.modeling_bert.BertLayer,
           model=model,
           config=config)
```

short-term TODO (not in this PR): add support for hashing the `orig_module_impl` to ensure we know how to translate between modules and they haven't changed.

File: deepspeed/ops/transformer/transformer.py (hunk 1 above)
Comment (tjruwase) [reply deepspeed/ops/transformer/transformer.py:411]: No, I did not observe any perf impact except for the memory savings.

File: deepspeed/ops/transformer/transformer.py (hunk 2 above)
Comment (tjruwase) [reply deepspeed/ops/transformer/transformer.py:254]: Thanks!

File: deepspeed/ops/transformer/transformer.py (hunk 1 above)
Comment (RezaYazdaniAminabadi) [reply deepspeed/ops/transformer/transformer.py:411]: this is great, thanks Tunji for investigating this :)

**Review by jeffra — APPROVED** [review]

**Pull request merged**